File: web/src/main/java/top/xuqingquan/web/system/JsBaseInterfaceHolder.java
package top.xuqingquan.web.system;

import android.support.annotation.NonNull;
import android.webkit.JavascriptInterface;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;

import top.xuqingquan.web.nokernel.JsInterfaceHolder;
import top.xuqingquan.web.nokernel.WebConfig;

@SuppressWarnings("rawtypes")
public abstract class JsBaseInterfaceHolder implements JsInterfaceHolder {

    private WebCreator mWebCreator;

    protected JsBaseInterfaceHolder(WebCreator webCreator) {
        this.mWebCreator = webCreator;
    }

    @Override
    public boolean checkObject(@NonNull Object v) {
        if (mWebCreator.getWebViewType() == WebConfig.WEBVIEW_AGENTWEB_SAFE_TYPE) {
            return true;
        }
        boolean tag = false;
        Class clazz = v.getClass();
        Method[] mMethods = clazz.getMethods();
        for (Method mMethod : mMethods) {
            Annotation[] mAnnotations = mMethod.getAnnotations();
            for (Annotation mAnnotation : mAnnotations) {
                if (mAnnotation instanceof JavascriptInterface) {
                    tag = true;
                    break;
                }
            }
            if (tag) {
                break;
            }
        }
        return tag;
    }

}

File: web/src/main/java/top/xuqingquan/web/system/JsBaseInterfaceHolder.kt
package top.xuqingquan.web.system

import android.webkit.JavascriptInterface
import top.xuqingquan.web.nokernel.JsInterfaceHolder
import top.xuqingquan.web.nokernel.WebConfig

/**
 * Created by 许清泉 on 12/21/20 11:09 PM
 */
abstract class JsBaseInterfaceHolder protected constructor(private val webCreator: WebCreator<*>) :
    JsInterfaceHolder {

    override fun checkObject(v: Any): Boolean {
        if (webCreator.getWebViewType() == WebConfig.WEB_VIEW_AGENT_WEB_SAFE_TYPE) {
            return true
        }
        var tag = false
        val clazz = v.javaClass
        val methods = clazz.methods
        for (method in methods) {
            val annotations = method.annotations
            for (annotation in annotations) {
                if (annotation is JavascriptInterface) {
                    tag = true
                    break
                }
            }
            if (tag){
                break
            }
        }
        return tag
    }

}